### ISSUE TITLE
AssemblyLoadContext.InitiateUnload - null argument on lock

### ISSUE THREAD

**Issue by TETYYS** [dotnet/coreclr#21100]
Just trying out ML.NET, `IEstimator.Fit` throws this:
```
System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at System.Threading.Monitor.ReliableEnter(Object obj, Boolean& lockTaken)
   at System.Runtime.Loader.AssemblyLoadContext.InitiateUnload()
   at System.Runtime.Loader.AssemblyLoadContext.Finalize()
```

`Microsoft.NETCore.App`: 3.0.0-preview-27117-01

@janvorli ?

**Comment by jkotas** [comment]
cc @janvorli

**Comment by janvorli** [comment]
@TETYYS thank you for reporting that. The InitiateUnload needs to be modified to behave differently for the case it is called from the finalizer (so that it doesn't access managed members of the AssemblyLoadContext, like the Unloading, unloadLock or ContextsToUnload).

**Comment by TETYYS** [comment]
Uhh, I am actually not sure myself that this is related to unloadability, it seems that the consequence of the bug is breaking unloadability. I was trying to reproduce one very weird bug with the compiler earlier that might be associated with this, but couldn't reproduce it without the entire project, I'll keep trying through.

**Comment by janvorli** [comment]
Actually, even without the GC.SuppressFinalize(), the finalizer should never get called. In such a case, we create a strong handle that should keep the AssemblyLoadContext alive. So I am really baffled here.
@TETYYS would it be possible to share just your AssemblyLoadContext derivation that your project uses?

**Comment by TETYYS** [comment]
The problem is that ML.NET uses that, not me. I think my project is doing something very weird that causes it to fail, I can't pinpoint what.

**Comment by janvorli** [comment]
I've searched the whole ml.net github and it doesn't do anything with AssemblyLoadContext at all. There is one other way how AssemblyLoadContext can get created - using Assembly.LoadFile. But ML.NET doesn't use that either. Do you happen to use it?
If not, then it would have to be the default AssemblyLoadContext.

**Comment by TETYYS** [comment]
https://github.com/dotnet/machinelearning/blob/d885ebc49c1fff04209714bb2757166a6f2681c9/src/Microsoft.ML.Core/ComponentModel/AssemblyLoadingUtils.cs#L211 ?

**Comment by janvorli** [comment]
Oh, right. I can see I had a typo when I was looking for the `LoadFrom` in the repo.
Anyways, as for memory corruption debugging, there is a couple of tools. You can set `COMPlus_HeapVerify` env variable to 1 and then GC heap consistency would be verified after each GC. That way, the corruption would likely be caught earlier.
Then you can use WinDBG with sos plugin. It also has a command to verify GC heap and you can also let windbg to break in when the ArgumentNullException is thrown and then dump the AssemblyLoadContext. You can use `!dso` SOS command to dump objects on stack and look for AssemblyLoadContext or you can use `!dumpheap -type AssemblyLoadContext` to dump all instances of classes that have the AssemblyLoadContext in their name.

**Comment by danmosemsft** [comment]
@TETYYS the first thing I suggest is to see whether it reproes on .NET Framework (and if not, try .NET Core 2.1). That will help indicate whether it is a runtime bug. If it reproes on .NET Framework it is almost surely an app issue.

**Comment by janvorli** [comment]
Should be fixed by #21189
